### ISSUE TITLE
Header exchange que bindings don't receive messages

### ISSUE THREAD

**Issue by matt01** [net-amqp-rabbitmq/net-amqp-rabbitmq#99]
When doing the following:
$mq->queue_bind(1, $queuename, $exchangename, '', {'header' => 'value'})

The binding is shown in the rabbitmq web interface with the custom arguments 'header: value', however, the que doesn't receive any messages from the exchange even know there are messages that match.

As a test, changing the line:
```
amqp_queue_bind(conn, channel, amqp_cstring_bytes(queuename),
                amqp_cstring_bytes(exchange),
                amqp_cstring_bytes(bindingkey),
                arguments);
```
to ...
```
    amqp_table_entry_t entries[1];
    amqp_table_t table;

    entries[0].key = amqp_cstring_bytes("header");
    entries[0].value.kind = AMQP_FIELD_KIND_UTF8;
    entries[0].value.value.bytes = amqp_cstring_bytes("value");

    table.num_entries = 1;
    table.entries = entries;

    amqp_queue_bind(conn, channel, amqp_cstring_bytes(queuename),
                    amqp_cstring_bytes(exchange),
                    amqp_cstring_bytes(bindingkey),
                    table);
```
in RabbitMQ.xs (net_amqp_rabbitmq_queue_bind) produces the expected results, the perl script runs and only sees the messages that match the header.

I'm thinking the encoding might be wrong for the header value but not sure.

**Comment by manchicken** [comment]
Your sample code doesn't seem to be a complete example... (the before has a
lot fewer lines than the after). Are you sure you meant it that way?

**Comment by manchicken** [comment]
Yeah, I can't make heads or tails of your example. Please include the full diff.

**Comment by matt01** [comment]
```
diff --git a/RabbitMQ.xs b/RabbitMQ.xs
index 058e6e7..f1e44bd 100644
--- a/RabbitMQ.xs
+++ b/RabbitMQ.xs
@@ -1300,12 +1300,23 @@ net_amqp_rabbitmq_queue_bind(conn, channel, queuename, exchange, bindingkey, arg
       Perl_croak(aTHX_ "queuename and exchange must both be specified");
     }
 
-    if(args)
-      hash_to_amqp_table(args, &arguments, 0);
+    //if(args)
+    //  hash_to_amqp_table(args, &arguments, 0);
+
+    amqp_table_entry_t entries[1];
+    amqp_table_t table;
+
+    entries[0].key = amqp_cstring_bytes("header");
+    entries[0].value.kind = AMQP_FIELD_KIND_UTF8;
+    entries[0].value.value.bytes = amqp_cstring_bytes("value");
+
+    table.num_entries = 1;
+    table.entries = entries;
+
     amqp_queue_bind(conn, channel, amqp_cstring_bytes(queuename),
                     amqp_cstring_bytes(exchange),
                     amqp_cstring_bytes(bindingkey),
-                    arguments);
+                    table);
     maybe_recycle_memory( conn );
     die_on_amqp_error(aTHX_ amqp_get_rpc_reply(conn), conn, "Binding queue");
```

**Comment by matt01** [comment]
This isn't a fix, its just a test as the header and value to match against is not being taken from perl. When the arguments to amqp_queue_bind are read from perl the rabbitmq web interface shows the headers but the que doesn't seem to receive any messages.

**Comment by matt01** [comment]
Ok, Just tried this ...

```
diff --git a/RabbitMQ.xs b/RabbitMQ.xs
index 058e6e7..554d2f6 100644
--- a/RabbitMQ.xs
+++ b/RabbitMQ.xs
@@ -1301,7 +1301,7 @@ net_amqp_rabbitmq_queue_bind(conn, channel, queuename, exchange, bindingkey, arg
     }
 
     if(args)
-      hash_to_amqp_table(args, &arguments, 0);
+      hash_to_amqp_table(args, &arguments, 1);
     amqp_queue_bind(conn, channel, amqp_cstring_bytes(queuename),
                     amqp_cstring_bytes(exchange),
                     amqp_cstring_bytes(bindingkey),

```
and it seems to be working for me. Looks like it might be related to utf8 encoding.

**Comment by manchicken** [comment]
Yeah, your strings may not be UTF8 magicked when they need to be. What's
the Perl data dumper of your arguments?

**Comment by manchicken** [comment]
That 1 just forces strings to UTF8 iirc

**Comment by ikegami** [comment]
I have reviewed the code in hash_to_amqp_table affected by its third argument. It's fine. (In fact, it might be the only place that handles magic!) This appears to be a user error, not a bug. e.g. The user's source code is encoded using UTF-8, but they didn't tell Perl that by using `use utf8;`.

**Comment by matt01** [comment]
Testing this more it works when sending messages from perl to another perl script, sorry I probably should of said the messages I am having this issue with are ones being generated by syslog-ng using its amqp() driver. This still persists if `use utf8;` is used or not.

I will try sending a message using rabbitmq-c and see if the same happens.

I have only seen this when receiving messages using net-amqp-rabbitmq. Receiving using python and C seems to 'just work'.

Thanks, 
Matt.

**Comment by matt01** [comment]
Also, in case it helps here is the config i'm using for syslog-ng:

```
destination d_amqp {
    amqp(
        vhost("/")
        host("127.0.0.1")
        port(5672)
        username("user")
        password("pass")
        exchange("exch.syslog")
        exchange-type("headers")
        body("$MSG")
        persistent(no)
        exchange-declare(yes)
        value-pairs(
            pair("syslog-host", "$HOST")
            pair("syslog-facility", "$FACILITY")
            pair("syslog-priority", "$PRIORITY")
            pair("syslog-level", "$LEVEL")
            pair("syslog-tag", "$TAG")
            pair("syslog-year", "$YEAR")
            pair("syslog-month", "$MONTH")
            pair("syslog-day", "$DAY")
            pair("syslog-hour", "$HOUR")
            pair("syslog-min", "$MIN")
            pair("syslog-sec", "$SEC")
            pair("syslog-program", "$PROGRAM")
        )
    );
};
```

**Comment by ikegami** [comment]
Everything you've just said is consistent with my earlier conclusion. Add the following before your call to queue_bind:

    use Data::Dumper;
    {
       local $Data::Dumper::Useqq = 1;
       local $Data::Dumper::Sortkeys = 1;
       warn(Dumper($arguments));
    }

**Comment by matt01** [comment]
Code:
```
my $header = 'testheader';
my $value = 'testvalue';

use Data::Dumper;
{
   local $Data::Dumper::Useqq = 1;
   local $Data::Dumper::Sortkeys = 1;

   warn(Dumper($header));
   warn(Dumper($value));
}

use Test::utf8;
is_sane_utf8($header);
is_sane_utf8($value);

$mq->queue_bind(1, $queuename, $exchangename, '', {$header => $value});
```
Output:
```
$VAR1 = "testheader";
$VAR1 = "testvalue";
ok 1 - sane utf8
ok 2 - sane utf8
```

**Comment by ikegami** [comment]
That's exactly what you used? wow. Maybe something is refusing to compare values of type AMQP_FIELD_KIND_UTF8 against values of type AMQP_FIELD_KIND_BYTES? If so, you can tell $mq how to encode it by using

    utf8::upgrade($_) for $header, $value;

**Comment by matt01** [comment]
Yea, that works, Humm...

**Comment by ikegami** [comment]
Sounds like something like force_utf8_in_header_strings is needed here too. That said, the force_utf8_in_header_strings code looks buggy to me. (Should use SvPVutf8 instead of SvPV, Otherwise, the strring "Éric" can result in junk (non-UTF8) depending other its initial storage format.)

**Comment by manchicken** [comment]
Pretty sure there's magic in Perl now that `utf8` is so built-in where all hard-coded strings are automatically encoded to UTF-8.

For `SvPVutf8()`, I think that's a great idea but I don't see it documented in `perlguts` at all. It's important to remember that undocumented Perl internals are subject to change without notice from version to version.

Also, we should absolutely not treat strings like UTF-8 by default. Perl already has conventions on how to deal with UTF-8, and while most strings can be transcoded from their original character set to UTF-8, not all of them can be, so we can't just assume internal storage is okay.

The convention of this library has always been that we handle UTF-8 if you already told Perl that you were using UTF-8 for the string. We don't assume or encode for you because we don't want to screw up your strings.

Some of the UTF-8 handling in the header is on account of some headers being UTF-8 per spec, thus making it our responsibility to make sure that those strings are encoded that way.

**Comment by ikegami** [comment]
> Pretty sure there's magic in Perl now that utf8 is so built-in where all hard-coded strings are automatically encoded to UTF-8.

Most definitely not. That's why utf8::upgrade had an effect.

```
$ perl -MDevel::Peek -e'Dump("abcde")' 2>&1 | grep FLAGS
  FLAGS = (PADTMP,POK,READONLY,pPOK)

$ perl -MDevel::Peek -e'use utf8; Dump("abcde")' 2>&1 | grep FLAGS
  FLAGS = (PADTMP,POK,READONLY,pPOK)

$ perl -MDevel::Peek -e'use utf8; Dump("abcdé")' 2>&1 | grep FLAGS
  FLAGS = (PADTMP,POK,READONLY,pPOK,UTF8)
```

**Comment by markwellis** [comment]
So this boils down to, one thing is using utf8, and another isn't

I could be wrong, but I don't think there's anything we can do to make this work nicer, without breaking things for other people.

Am I understanding this right? Encodings are confusing!

**Comment by ikegami** [comment]
It boils down to: A value of type AMQP_FIELD_KIND_UTF8 is never equal to a value of type AMQP_FIELD_KIND_BYTES, and NAR uses the internal storage format to decided on the type of the value.

I agree that the current behaviour should be maintained by default, but it would be possible for queue_bind to accept an optional "options" hash argument, in which force_utf8_in_header_strings could be specified.

**Comment by manchicken** [comment]
perl -e is not hard-coded.

**Comment by ikegami** [comment]
@manchicken, Are you implying `perl` might treat code from `-e` differently than code from a file? It doesn't. Feel free to test it.

**Comment by manchicken** [comment]
Okay, I'm going to dig into this fella. I like @ikegami's suggestion, and I can't remember a lick of the context that was in my brain when we ended this conversation before.
